File: YouTubePlayer/src/main/java/com/pierfrancescosoffritti/youtubeplayer/ui/DefaultPlayerUIController.java
package com.pierfrancescosoffritti.youtubeplayer.ui;

import android.animation.Animator;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;

import com.pierfrancescosoffritti.youtubeplayer.R;
import com.pierfrancescosoffritti.youtubeplayer.YouTubePlayer;
import com.pierfrancescosoffritti.youtubeplayer.YouTubePlayerFullScreenListener;
import com.pierfrancescosoffritti.youtubeplayer.YouTubePlayerView;
import com.pierfrancescosoffritti.youtubeplayer.utils.Utils;

public class DefaultPlayerUIController implements PlayerUIController, View.OnClickListener, YouTubePlayerFullScreenListener, YouTubePlayer.YouTubePlayerListener, SeekBar.OnSeekBarChangeListener {
    @NonNull private final YouTubePlayerView youTubePlayerView;

    // view responsible for intercepting clicks. Could have used controlsRoot view, but in this way I'm able to hide all the control at once by hiding controlsRoot
    @NonNull private final View panel;

    // view containing the controls
    @NonNull private final View controlsRoot;

    @NonNull private final TextView videoTitle;
    @NonNull private final TextView videoCurrentTime;
    @NonNull private final TextView videoDuration;

    @NonNull private final ProgressBar progressBar;
    @NonNull private final ImageView playButton;
    @NonNull private final ImageView youTubeButton;
    @NonNull private final ImageView fullScreenButton;

    @NonNull private final ImageView customActionLeft;
    @NonNull private final ImageView customActionRight;

    @NonNull private final SeekBar seekBar;

    private View.OnClickListener onFullScreenButtonListener;

    // view state
    private boolean isPlaying = false;
    private boolean isVisible = true;
    private boolean canFadeControls = false;

    private boolean showUI = true;

    public DefaultPlayerUIController(@NonNull YouTubePlayerView youTubePlayerView, @NonNull View controlsView) {
        this.youTubePlayerView = youTubePlayerView;

        panel = controlsView.findViewById(R.id.panel);

        controlsRoot = controlsView.findViewById(R.id.controls_root);

        videoTitle = controlsView.findViewById(R.id.video_title);
        videoCurrentTime = controlsView.findViewById(R.id.video_current_time);
        videoDuration = controlsView.findViewById(R.id.video_duration);

        progressBar = controlsView.findViewById(R.id.progress);
        playButton = controlsView.findViewById(R.id.play_button);
        youTubeButton = controlsView.findViewById(R.id.youtube_button);
        fullScreenButton = controlsView.findViewById(R.id.fullscreen_button);

        customActionLeft = controlsView.findViewById(R.id.custom_action_left_button);
        customActionRight = controlsView.findViewById(R.id.custom_action_right_button);

        seekBar = controlsView.findViewById(R.id.seek_bar);

        seekBar.setOnSeekBarChangeListener(this);
        panel.setOnClickListener(this);
        playButton.setOnClickListener(this);
        fullScreenButton.setOnClickListener(this);
    }

    @Override
    public void showTitle(boolean show) {
        videoTitle.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override
    public void showUI(boolean show) {
        if(show)
            controlsRoot.setVisibility(View.VISIBLE);
        else
            controlsRoot.setVisibility(View.INVISIBLE);

        showUI = show;
    }

    /**
     * Set custom action to the left of the Play/Pause button
     */
    @Override
    public void setCustomAction1(Drawable icon, View.OnClickListener clickListener) {
        customActionLeft.setImageDrawable(icon);
        customActionLeft.setOnClickListener(clickListener);
        showCustomAction1(clickListener != null);
    }

    /**
     * Set custom action to the right of the Play/Pause button
     */
    @Override
    public void setCustomAction2(Drawable icon, View.OnClickListener clickListener) {
        customActionRight.setImageDrawable(icon);
        customActionRight.setOnClickListener(clickListener);
        showCustomAction2(clickListener != null);
    }

    public void showCustomAction1(boolean show) {
        int visibility = show ? View.VISIBLE : View.GONE;
        customActionLeft.setVisibility(visibility);
    }

    public void showCustomAction2(boolean show) {
        int visibility = show ? View.VISIBLE : View.GONE;
        customActionRight.setVisibility(visibility);
    }

    @Override
    public void showFullscreenButton(boolean show) {
        int visibility = show ? View.VISIBLE : View.INVISIBLE;
        fullScreenButton.setVisibility(visibility);
    }

    @Override
    public void setCustomFullScreenButtonListener(View.OnClickListener customFullScreenButtonListener) {
        this.onFullScreenButtonListener = customFullScreenButtonListener;
    }

    @Override
    public void onClick(View view) {
        if(view == panel)
            toggleControlsVisibility();
        else if(view == playButton)
            onPlayButtonPressed();
        else if(view == fullScreenButton)
            onFullScreenPressed();
    }

    private void onFullScreenPressed() {
        if(onFullScreenButtonListener == null)
            youTubePlayerView.toggleFullScreen();
        else
            onFullScreenButtonListener.onClick(fullScreenButton);
    }

    private void onPlayButtonPressed() {
        if(isPlaying)
            youTubePlayerView.pause();
        else
            youTubePlayerView.play();
    }

    private void updatePlayPauseButtonIcon(boolean playing) {
        int img = playing ? R.drawable.ic_pause_36dp : R.drawable.ic_play_36dp;
        playButton.setImageResource(img);
    }

    private void toggleControlsVisibility() {
        final float finalAlpha = isVisible ? 0f : 1f;
        fadeControls(finalAlpha);
    }

    private final Handler handler = new Handler(Looper.getMainLooper());
    private final Runnable fadeOutRunnable = new Runnable() {
        @Override
        public void run() {
            fadeControls(0f);
        }
    };

    private void fadeControls(final float finalAlpha) {
        if(!canFadeControls || !showUI)
            return;

        isVisible = finalAlpha != 0f;

        // if the controls are shown and the player is playing they should automatically hide after a while.
        // if the controls are hidden remove fade out runnable
        if(finalAlpha == 1f && isPlaying)
            startFadeOutViewTimer();
        else
            handler.removeCallbacks(fadeOutRunnable);


        controlsRoot.animate()
                .alpha(finalAlpha)
                .setDuration(300)
                .setListener(new Animator.AnimatorListener() {
                    @Override
                    public void onAnimationStart(Animator animator) {
                        if (finalAlpha == 1f)
                            controlsRoot.setVisibility(View.VISIBLE);
                    }

                    @Override
                    public void onAnimationEnd(Animator animator) {
                        if (finalAlpha == 0f)
                            controlsRoot.setVisibility(View.GONE);
                    }

                    @Override public void onAnimationCancel(Animator animator) { }
                    @Override public void onAnimationRepeat(Animator animator) { }
                }).start();
    }

    private void startFadeOutViewTimer() {
        handler.postDelayed(fadeOutRunnable, 3000);
    }

    @Override
    public void onYouTubePlayerEnterFullScreen() {
        fullScreenButton.setImageResource(R.drawable.ic_fullscreen_exit_24dp);
    }

    @Override
    public void onYouTubePlayerExitFullScreen() {
        fullScreenButton.setImageResource(R.drawable.ic_fullscreen_24dp);
    }

    // YouTubePlayer callbacks

    // TODO refactor this method
    @Override
    public void onStateChange(@YouTubePlayer.PlayerState.State int state) {
        System.out.println("state: " +state);
        newSeekBarProgress = -1;

        updateControlsState(state);

        if(state == YouTubePlayer.PlayerState.PLAYING || state == YouTubePlayer.PlayerState.PAUSED || state == YouTubePlayer.PlayerState.VIDEO_CUED) {
            panel.setBackgroundColor(ContextCompat.getColor(youTubePlayerView.getContext(), android.R.color.transparent));
            progressBar.setVisibility(View.GONE);
            playButton.setVisibility(View.VISIBLE);


            canFadeControls = true;
            boolean playing = state == YouTubePlayer.PlayerState.PLAYING;
            updatePlayPauseButtonIcon(playing);

            if(playing)
                startFadeOutViewTimer();
            else
                handler.removeCallbacks(fadeOutRunnable);

        } else {
            updatePlayPauseButtonIcon(false);
            fadeControls(1f);

            if(state == YouTubePlayer.PlayerState.BUFFERING) {
                playButton.setVisibility(View.INVISIBLE);

                customActionLeft.setVisibility(View.GONE);
                customActionRight.setVisibility(View.GONE);

                progressBar.setVisibility(View.VISIBLE);
                canFadeControls = false;
            }

            if(state == YouTubePlayer.PlayerState.UNSTARTED) {
                panel.setBackgroundColor(ContextCompat.getColor(youTubePlayerView.getContext(), android.R.color.black));
                canFadeControls = false;

                progressBar.setVisibility(View.GONE);
                playButton.setVisibility(View.VISIBLE);
            }
        }
    }

    private void updateControlsState(int state) {
        switch (state) {
            case YouTubePlayer.PlayerState.ENDED:
                isPlaying = false;
                break;
            case YouTubePlayer.PlayerState.PAUSED:
                isPlaying = false;
                break;
            case YouTubePlayer.PlayerState.PLAYING:
                isPlaying = true;
                break;
            case YouTubePlayer.PlayerState.UNSTARTED:
                resetUI();
                break;
            default:
                break;
        }


        updatePlayPauseButtonIcon(!isPlaying);
    }

    @Override
    public void onCurrentSecond(float second) {
        // ignore if the user is currently moving the SeekBar
        if(seekBarTouchStarted)
            return;
        // ignore if the current time is older than what the user selected with the SeekBar
        if(newSeekBarProgress > 0 && !Utils.formatTime(second).equals(Utils.formatTime(newSeekBarProgress)))
            return;

        newSeekBarProgress = -1;

        seekBar.setProgress((int) second);
    }

    @Override
    public void onVideoDuration(float duration) {
        videoDuration.setText(Utils.formatTime(duration));
        seekBar.setMax((int) duration);
    }

    @Override
    public void onVideoTitle(String title) {
        videoTitle.setText(title);
    }

    @Override
    public void onVideoId(final String videoId) {
        youTubeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("http://www.youtube.com/watch?v=" + videoId));
                controlsRoot.getContext().startActivity(intent);
            }
        });
    }

    @Override public void onReady() { }
    @Override public void onMessage(String log) { }
    @Override public void onPlaybackQualityChange(@YouTubePlayer.PlaybackQuality.Quality int playbackQuality) { }
    @Override public void onPlaybackRateChange(@YouTubePlayer.PlaybackRate.Rate String rate) { }
    @Override public void onError(@YouTubePlayer.PlayerError.Error int error) { }
    @Override public void onApiChange() { }

    // SeekBar callbacks

    private boolean seekBarTouchStarted = false;
    // I need this variable because onCurrentSecond gets called every 100 mils, so without the proper checks on this variable in onCurrentSeconds the seek bar glitches when touched.
    private int newSeekBarProgress = -1;

    @Override
    public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
        videoCurrentTime.setText(Utils.formatTime(i));
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        seekBarTouchStarted = true;
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        if(isPlaying)
            newSeekBarProgress = seekBar.getProgress();

        youTubePlayerView.seekTo(seekBar.getProgress());
        seekBarTouchStarted = false;
    }

    public void resetUI() {
        seekBar.setProgress(0);
        seekBar.setMax(0);

        videoDuration.post(new Runnable() {
            @Override
            public void run() {
                videoDuration.setText("");
            }
        });

        videoTitle.post(new Runnable() {
            @Override
            public void run() {
                videoTitle.setText("");
            }
        });

        youTubeButton.setOnClickListener(null);
    }
}

File: custom-ui/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/core/customui/DefaultPlayerUiController.kt
package com.pierfrancescosoffritti.androidyoutubeplayer.core.customui

import android.content.Intent
import android.graphics.drawable.Drawable
import android.net.Uri
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.PlayerConstants
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.YouTubePlayer
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.listeners.AbstractYouTubePlayerListener
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.views.YouTubePlayerView
import com.pierfrancescosoffritti.androidyoutubeplayer.core.customui.menu.YouTubePlayerMenu
import com.pierfrancescosoffritti.androidyoutubeplayer.core.customui.menu.defaultMenu.DefaultYouTubePlayerMenu
import com.pierfrancescosoffritti.androidyoutubeplayer.core.customui.utils.FadeViewHelper
import com.pierfrancescosoffritti.androidyoutubeplayer.core.customui.views.YouTubePlayerSeekBar
import com.pierfrancescosoffritti.androidyoutubeplayer.core.customui.views.YouTubePlayerSeekBarListener

class DefaultPlayerUiController(
  private val youTubePlayerView: YouTubePlayerView,
  private val youTubePlayer: YouTubePlayer
) : PlayerUiController {

  val rootView: View = View.inflate(youTubePlayerView.context, R.layout.ayp_default_player_ui, null)

  private var youTubePlayerMenu: YouTubePlayerMenu = DefaultYouTubePlayerMenu(
    youTubePlayerView.context
  )

  /**
   * View used for for intercepting clicks and for drawing a black background.
   * Could have used controlsContainer, but in this way I'm able to hide all the control at once by hiding controlsContainer
   */
  private val panel: View = rootView.findViewById(R.id.panel)

  private val controlsContainer: View = rootView.findViewById(R.id.controls_container)
  private val extraViewsContainer: LinearLayout = rootView.findViewById(R.id.extra_views_container)

  private val videoTitle: TextView = rootView.findViewById(R.id.video_title)
  private val liveVideoIndicator: TextView = rootView.findViewById(R.id.live_video_indicator)

  private val progressBar: ProgressBar = rootView.findViewById(R.id.progress)
  private val menuButton: ImageView = rootView.findViewById(R.id.menu_button)
  private val playPauseButton: ImageView = rootView.findViewById(R.id.play_pause_button)
  private val youTubeButton: ImageView = rootView.findViewById(R.id.youtube_button)
  private val fullscreenButton: ImageView = rootView.findViewById(R.id.fullscreen_button)

  private val customActionLeft: ImageView = rootView.findViewById(R.id.custom_action_left_button)
  private val customActionRight: ImageView = rootView.findViewById(R.id.custom_action_right_button)

  private val youtubePlayerSeekBar: YouTubePlayerSeekBar = rootView.findViewById(R.id.youtube_player_seekbar)
  private val fadeControlsContainer: FadeViewHelper = FadeViewHelper(controlsContainer)

  private var onFullscreenButtonListener: View.OnClickListener
  private var onMenuButtonClickListener: View.OnClickListener

  private var isPlaying = false
  private var isPlayPauseButtonEnabled = true
  private var isCustomActionLeftEnabled = false
  private var isCustomActionRightEnabled = false

  private var isMatchParent = false

  private val youTubePlayerStateListener = object : AbstractYouTubePlayerListener() {
    override fun onStateChange(youTubePlayer: YouTubePlayer, state: PlayerConstants.PlayerState) {
      updateState(state)

      if (state === PlayerConstants.PlayerState.PLAYING || state === PlayerConstants.PlayerState.PAUSED || state === PlayerConstants.PlayerState.VIDEO_CUED) {
        panel.setBackgroundColor(ContextCompat.getColor(panel.context, android.R.color.transparent))
        progressBar.visibility = View.GONE

        if (isPlayPauseButtonEnabled) playPauseButton.visibility = View.VISIBLE
        if (isCustomActionLeftEnabled) customActionLeft.visibility = View.VISIBLE
        if (isCustomActionRightEnabled) customActionRight.visibility = View.VISIBLE

        updatePlayPauseButtonIcon(state === PlayerConstants.PlayerState.PLAYING)

      } else {
        updatePlayPauseButtonIcon(false)

        if (state === PlayerConstants.PlayerState.BUFFERING) {
          progressBar.visibility = View.VISIBLE
          panel.setBackgroundColor(
            ContextCompat.getColor(
              panel.context,
              android.R.color.transparent
            )
          )
          if (isPlayPauseButtonEnabled) playPauseButton.visibility = View.INVISIBLE

          customActionLeft.visibility = View.GONE
          customActionRight.visibility = View.GONE
        }

        if (state === PlayerConstants.PlayerState.UNSTARTED) {
          progressBar.visibility = View.GONE
          if (isPlayPauseButtonEnabled) playPauseButton.visibility = View.VISIBLE
        }
      }
    }

    override fun onVideoId(youTubePlayer: YouTubePlayer, videoId: String) {
      youTubeButton.setOnClickListener {
        val intent = Intent(
          Intent.ACTION_VIEW,
          Uri.parse("https://www.youtube.com/watch?v=" + videoId + "#t=" + youtubePlayerSeekBar.seekBar.progress)
        )
        try {
          youTubeButton.context.startActivity(intent)
        } catch (e: Exception) {
          Log.e(javaClass.simpleName, e.message ?: "Can't open url to YouTube")
        }
      }
    }
  }

  init {
    onFullscreenButtonListener = View.OnClickListener {
      isMatchParent = !isMatchParent
      when (isMatchParent) {
        true -> youTubePlayerView.matchParent()
        false -> youTubePlayerView.wrapContent()
      }
    }

    onMenuButtonClickListener = View.OnClickListener { youTubePlayerMenu.show(menuButton) }

    initClickListeners()
  }

  private fun initClickListeners() {
    youTubePlayer.addListener(youtubePlayerSeekBar)
    youTubePlayer.addListener(fadeControlsContainer)
    youTubePlayer.addListener(youTubePlayerStateListener)

    youtubePlayerSeekBar.youtubePlayerSeekBarListener = object : YouTubePlayerSeekBarListener {
      override fun seekTo(time: Float) = youTubePlayer.seekTo(time)
    }
    panel.setOnClickListener { fadeControlsContainer.toggleVisibility() }
    playPauseButton.setOnClickListener { onPlayButtonPressed() }
    fullscreenButton.setOnClickListener { onFullscreenButtonListener.onClick(fullscreenButton) }
    menuButton.setOnClickListener { onMenuButtonClickListener.onClick(menuButton) }
  }

  override fun showVideoTitle(show: Boolean): PlayerUiController {
    videoTitle.visibility = if (show) View.VISIBLE else View.GONE
    return this
  }

  override fun setVideoTitle(videoTitle: String): PlayerUiController {
    this.videoTitle.text = videoTitle
    return this
  }

  override fun showUi(show: Boolean): PlayerUiController {
    fadeControlsContainer.isDisabled = !show
    controlsContainer.visibility = if (show) View.VISIBLE else View.INVISIBLE
    return this
  }

  override fun showPlayPauseButton(show: Boolean): PlayerUiController {
    playPauseButton.visibility = if (show) View.VISIBLE else View.GONE

    isPlayPauseButtonEnabled = show
    return this
  }

  override fun enableLiveVideoUi(enable: Boolean): PlayerUiController {
    youtubePlayerSeekBar.visibility = if (enable) View.INVISIBLE else View.VISIBLE
    liveVideoIndicator.visibility = if (enable) View.VISIBLE else View.GONE
    return this
  }

  override fun setCustomAction1(
    icon: Drawable,
    clickListener: View.OnClickListener?
  ): PlayerUiController {
    customActionLeft.setImageDrawable(icon)
    customActionLeft.setOnClickListener(clickListener)
    showCustomAction1(true)
    return this
  }

  override fun setCustomAction2(
    icon: Drawable,
    clickListener: View.OnClickListener?
  ): PlayerUiController {
    customActionRight.setImageDrawable(icon)
    customActionRight.setOnClickListener(clickListener)
    showCustomAction2(true)
    return this
  }

  override fun showCustomAction1(show: Boolean): PlayerUiController {
    isCustomActionLeftEnabled = show
    customActionLeft.visibility = if (show) View.VISIBLE else View.GONE
    return this
  }

  override fun showCustomAction2(show: Boolean): PlayerUiController {
    isCustomActionRightEnabled = show
    customActionRight.visibility = if (show) View.VISIBLE else View.GONE
    return this
  }

  override fun showMenuButton(show: Boolean): PlayerUiController {
    menuButton.visibility = if (show) View.VISIBLE else View.GONE
    return this
  }

  override fun setMenuButtonClickListener(customMenuButtonClickListener: View.OnClickListener): PlayerUiController {
    onMenuButtonClickListener = customMenuButtonClickListener
    return this
  }

  override fun showCurrentTime(show: Boolean): PlayerUiController {
    youtubePlayerSeekBar.videoCurrentTimeTextView.visibility = if (show) View.VISIBLE else View.GONE
    return this
  }

  override fun showDuration(show: Boolean): PlayerUiController {
    youtubePlayerSeekBar.videoDurationTextView.visibility = if (show) View.VISIBLE else View.GONE
    return this
  }

  override fun showSeekBar(show: Boolean): PlayerUiController {
    youtubePlayerSeekBar.seekBar.visibility = if (show) View.VISIBLE else View.INVISIBLE
    return this
  }

  override fun showBufferingProgress(show: Boolean): PlayerUiController {
    youtubePlayerSeekBar.showBufferingProgress = show
    return this
  }

  override fun showYouTubeButton(show: Boolean): PlayerUiController {
    youTubeButton.visibility = if (show) View.VISIBLE else View.GONE
    return this
  }

  override fun addView(view: View): PlayerUiController {
    extraViewsContainer.addView(view, 0)
    return this
  }

  override fun removeView(view: View): PlayerUiController {
    extraViewsContainer.removeView(view)
    return this
  }

  override fun getMenu(): YouTubePlayerMenu = youTubePlayerMenu

  override fun showFullscreenButton(show: Boolean): PlayerUiController {
    fullscreenButton.visibility = if (show) View.VISIBLE else View.GONE
    return this
  }

  override fun setFullscreenButtonClickListener(customFullscreenButtonClickListener: View.OnClickListener): PlayerUiController {
    onFullscreenButtonListener = customFullscreenButtonClickListener
    return this
  }

  private fun onPlayButtonPressed() {
    if (isPlaying)
      youTubePlayer.pause()
    else
      youTubePlayer.play()
  }

  private fun updateState(state: PlayerConstants.PlayerState) {
    when (state) {
      PlayerConstants.PlayerState.ENDED -> isPlaying = false
      PlayerConstants.PlayerState.PAUSED -> isPlaying = false
      PlayerConstants.PlayerState.PLAYING -> isPlaying = true
      else -> {}
    }

    updatePlayPauseButtonIcon(!isPlaying)
  }

  private fun updatePlayPauseButtonIcon(playing: Boolean) {
    val drawable = if (playing) R.drawable.ayp_ic_pause_36dp else R.drawable.ayp_ic_play_36dp
    playPauseButton.setImageResource(drawable)
  }
}
